File: Daily Questions/0670-maximum-swap/0670-maximum-swap.py
class Solution:
    def maximumSwap(self, num: int) -> int:
        """
        Returns the maximum number by swapping at most two digits of the given integer.

        Parameters:
        num (int): The input integer.

        Returns:
        int: The maximum number after at most one swap.
        """
        # Convert the number to a list of digits
        digits = list(str(num))
        
        # Create a dictionary to track the last position of each digit
        last = {int(d): i for i, d in enumerate(digits)}
        
        # Traverse the digits and find the first smaller digit that can be swapped with a larger one
        for i, d in enumerate(digits):
            # Check if there is a larger digit after the current one
            for bigger in range(9, int(d), -1):
                if last.get(bigger, -1) > i:
                    # Swap the digits
                    digits[i], digits[last[bigger]] = digits[last[bigger]], digits[i]
                    # Return the new number
                    return int("".join(digits))
        
        # If no swap has been made, return the original number
        return num

File: 0670-maximum-swap/0670-maximum-swap.c
#include <stdio.h>
#include <string.h>

/**
 * @brief Returns the maximum number by swapping at most two digits of the given integer.
 *
 * @param num (int): The input integer.
 *
 * @return (int): The maximum number after at most one swap.
 */ 
int maximumSwap(int num) {
    char digits[10];
    sprintf(digits, "%d", num);  // Convert the number to a string for easy manipulation
    int n = strlen(digits);
    
    int last[10] = {0};
    
    // Record the last occurrence of each digit
    for (int i = 0; i < n; i++) {
        last[digits[i] - '0'] = i;
    }

    // Traverse the digits and try to find the best swap
    for (int i = 0; i < n; i++) {
        for (int d = 9; d > digits[i] - '0'; d--) {
            if (last[d] > i) {
                // Swap the digits
                char temp = digits[i];
                digits[i] = digits[last[d]];
                digits[last[d]] = temp;
                
                // Convert back to integer and return
                int swappedNum;
                sscanf(digits, "%d", &swappedNum);
                return swappedNum;
            }
        }
    }
    
    // Return the original number if no swap happened
    return num;
}
